#Fix proper invalidation for elements with absolute position
# need bug for mozilaa
diff -r cb59ce4e65e4 layout/generic/nsGfxScrollFrame.cpp
--- a/layout/generic/nsGfxScrollFrame.cpp	Mon Jun 15 11:49:56 2009 +0300
+++ b/layout/generic/nsGfxScrollFrame.cpp	Mon Jun 15 15:10:15 2009 +0300
@@ -1277,6 +1277,7 @@ nsGfxScrollFrameInner::nsGfxScrollFrameI
     mOuter(aOuter),
     mRestoreRect(-1, -1, -1, -1),
     mLastPos(-1, -1),
+    mWillChangePos(0, 0),
     mNeverHasVerticalScrollbar(PR_FALSE),
     mNeverHasHorizontalScrollbar(PR_FALSE),
     mHasVerticalScrollbar(PR_FALSE), 
@@ -1752,7 +1753,9 @@ nsGfxScrollFrameInner::Destroy()
 NS_IMETHODIMP
 nsGfxScrollFrameInner::ScrollPositionWillChange(nsIScrollableView* aScrollable, nscoord aX, nscoord aY)
 {
-   // Do nothing.
+   nsIScrollableView* s = GetScrollableView();
+   if (s)
+      s->GetScrollPosition(mWillChangePos.x, mWillChangePos.y);
    return NS_OK;
 }
 
@@ -1797,8 +1800,16 @@ nsGfxScrollFrameInner::ScrollPositionDid
   PostScrollEvent();
 
   // Notify that the display has changed
-  mOuter->InvalidateWithFlags(nsRect(nsPoint(0, 0), mOuter->GetSize()),
-                              nsIFrame::INVALIDATE_NOTIFY_ONLY);
+  nsRect r(nsPoint(0, 0), mOuter->GetSize());
+
+  // Area from previous scrollPort view also should be invalidated
+  nsPoint tr(aX - mWillChangePos.x, aY - mWillChangePos.y);
+  if (tr.y > 0) r.y -= tr.y;
+  if (tr.x > 0) r.x -= tr.x;
+  r.height += PR_ABS(tr.y);
+  r.width += PR_ABS(tr.x);
+
+  mOuter->InvalidateWithFlags(r, nsIFrame::INVALIDATE_NOTIFY_ONLY);
   return NS_OK;
 }
 
diff -r cb59ce4e65e4 layout/generic/nsGfxScrollFrame.h
--- a/layout/generic/nsGfxScrollFrame.h	Mon Jun 15 11:49:56 2009 +0300
+++ b/layout/generic/nsGfxScrollFrame.h	Mon Jun 15 15:10:15 2009 +0300
@@ -197,6 +197,7 @@ public:
 
   nsRect mRestoreRect;
   nsPoint mLastPos;
+  nsPoint mWillChangePos;
 
   PRPackedBool mNeverHasVerticalScrollbar:1;
   PRPackedBool mNeverHasHorizontalScrollbar:1;
